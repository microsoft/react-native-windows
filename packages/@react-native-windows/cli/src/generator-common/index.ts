/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 * @format
 */

import fs from 'fs';
import chalk from 'chalk';
import prompts from 'prompts';
import path from 'path';
import mustache from 'mustache';
import {CodedError} from '@react-native-windows/telemetry';

/**
 * Text to replace, + config options
 */
export type Replacements = {
  useMustache?: boolean;
  regExpPatternsToRemove?: RegExp[];
  [key: string]: any;
};

function walk(current: string): string[] {
  if (!fs.lstatSync(current).isDirectory()) {
    return [current];
  }

  const files = fs
    .readdirSync(current)
    .map(child => walk(path.join(current, child)));
  const result: string[] = [];
  return result.concat.apply([current], files);
}

/**
 * Get a source file and replace parts of its contents.
 * @param srcPath Path to the source file.
 * @param replacements e.g. {'TextToBeReplaced': 'Replacement'}
 * @return The contents of the file with the replacements applied.
 */
export function resolveContents(
  srcPath: string,
  replacements: Replacements,
): string {
  let content = fs.readFileSync(srcPath, 'utf8');

  if (content.includes('\r\n')) {
    // CRLF file, make sure multiline replacements are also CRLF
    for (const key of Object.keys(replacements)) {
      if (typeof replacements[key] === 'string') {
        replacements[key] = replacements[key].replace(/(?<!\r)\n/g, '\r\n');
      }
    }
  } else {
    // LF file, make sure multiline replacements are also LF
    for (const key of Object.keys(replacements)) {
      if (typeof replacements[key] === 'string') {
        replacements[key] = replacements[key].replace(/\r\n/g, '\n');
      }
    }
  }

  if (replacements.useMustache) {
    content = mustache.render(content, replacements);
    (replacements.regExpPatternsToRemove || []).forEach(regexPattern => {
      content = content.replace(new RegExp(regexPattern, 'g'), '');
    });
  } else {
    Object.keys(replacements).forEach(regex => {
      content = content.replace(new RegExp(regex, 'g'), replacements[regex]);
    });
  }
  return content;
}

// Binary files, don't process these (avoid decoding as utf8)
const binaryExtensions = ['.png', '.jar', '.keystore'];

type ContentChangedCallbackOption = 'identical' | 'changed' | 'new' | null;

type ContentChangedCallback = (
  path: string,
  option: ContentChangedCallbackOption,
) => Promise<'keep' | 'overwrite'>;

/**
 * Copy a file to given destination, replacing parts of its contents.
 * @param srcPath Path to a file to be copied.
 * @param destPath Destination path.
 * @param replacements: e.g. {'TextToBeReplaced': 'Replacement'}
 * @param contentChangedCallback
 *        Used when upgrading projects. Based on if file contents would change
 *        when being replaced, allows the caller to specify whether the file
 *        should be replaced or not.
 *        If null, files will be overwritten.
 *        Function(path, 'identical' | 'changed' | 'new') => 'keep' | 'overwrite'
 */
export async function copyAndReplace(
  srcPath: string,
  destPath: string,
  replacements: Replacements,
  contentChangedCallback: ContentChangedCallback | null,
) {
  if (fs.lstatSync(srcPath).isDirectory()) {
    if (!fs.existsSync(destPath)) {
      fs.mkdirSync(destPath);
    }
    // Not recursive
    return;
  }

  const extension = path.extname(srcPath);
  if (binaryExtensions.includes(extension)) {
    // Binary file
    let shouldOverwrite = 'overwrite';
    if (contentChangedCallback) {
      const newContentBuffer = fs.readFileSync(srcPath);
      let contentChanged: ContentChangedCallbackOption = 'identical';
      try {
        const origContentBuffer = fs.readFileSync(destPath);
        if (Buffer.compare(origContentBuffer, newContentBuffer) !== 0) {
          contentChanged = 'changed';
        }
      } catch (err) {
        if (err.code === 'ENOENT') {
          contentChanged = 'new';
        } else {
          throw err;
        }
      }
      shouldOverwrite = await contentChangedCallback(destPath, contentChanged);
    }
    if (shouldOverwrite === 'overwrite') {
      copyBinaryFile(srcPath, destPath, err => {
        if (err) {
          throw err;
        }
      });
    }
  } else {
    // Text file
    const srcPermissions = fs.statSync(srcPath).mode;
    const content = resolveContents(srcPath, replacements);

    let shouldOverwrite = 'overwrite';
    if (contentChangedCallback) {
      // Check if contents changed and ask to overwrite
      let contentChanged: ContentChangedCallbackOption = 'identical';
      try {
        const origContent = fs.readFileSync(destPath, 'utf8');
        if (content !== origContent) {
          // logger.info('Content changed: ' + destPath);
          contentChanged = 'changed';
        }
      } catch (err) {
        if (err.code === 'ENOENT') {
          contentChanged = 'new';
        } else {
          throw err;
        }
      }
      shouldOverwrite = await contentChangedCallback(destPath, contentChanged);
    }
    if (shouldOverwrite === 'overwrite') {
      fs.writeFileSync(destPath, content, {
        encoding: 'utf8',
        mode: srcPermissions,
      });
    }
  }
}

/**
 * Same as 'cp' on Unix. Don't do any replacements.
 */
function copyBinaryFile(
  srcPath: string,
  destPath: string,
  cb: (err?: Error) => void,
) {
  let cbCalled = false;
  const srcPermissions = fs.statSync(srcPath).mode;
  const readStream = fs.createReadStream(srcPath);
  readStream.on('error', err => {
    done(err);
  });
  const writeStream = fs.createWriteStream(destPath, {
    mode: srcPermissions,
  });
  writeStream.on('error', err => {
    done(err);
  });
  writeStream.on('close', () => {
    done();
  });
  readStream.pipe(writeStream);
  function done(err?: Error) {
    if (!cbCalled) {
      cb(err);
      cbCalled = true;
    }
  }
}

export function createDir(destPath: string) {
  if (!fs.existsSync(destPath)) {
    fs.mkdirSync(destPath);
  }
}

export async function copyAndReplaceWithChangedCallback(
  srcPath: string,
  destRoot: string,
  relativeDestPath: string,
  replacements?: Record<string, string>,
  alwaysOverwrite?: boolean,
) {
  if (!replacements) {
    replacements = {};
  }
  const contentChangedCallback: ContentChangedCallback = alwaysOverwrite
    ? (_, contentChanged) =>
        alwaysOverwriteContentChangedCallback(
          srcPath,
          relativeDestPath,
          contentChanged,
        )
    : (_, contentChanged) =>
        upgradeFileContentChangedCallback(
          srcPath,
          relativeDestPath,
          contentChanged,
        );

  await copyAndReplace(
    srcPath,
    path.join(destRoot, relativeDestPath),
    replacements,
    contentChangedCallback,
  );
}

export async function copyAndReplaceAll(
  srcPath: string,
  destPath: string,
  relativeDestDir: string,
  replacements: Replacements,
  alwaysOverwrite: boolean,
) {
  for (const absoluteSrcFilePath of walk(srcPath)) {
    const filename = path.relative(srcPath, absoluteSrcFilePath);
    const relativeDestPath = path.join(relativeDestDir, filename);
    await copyAndReplaceWithChangedCallback(
      absoluteSrcFilePath,
      destPath,
      relativeDestPath,
      replacements,
      alwaysOverwrite,
    );
  }
}

async function alwaysOverwriteContentChangedCallback(
  absoluteSrcFilePath: string,
  relativeDestPath: string,
  contentChanged: ContentChangedCallbackOption,
): Promise<'keep' | 'overwrite'> {
  if (contentChanged === 'new') {
    console.log(`${chalk.bold('new')} ${relativeDestPath}`);
    return 'overwrite';
  }
  if (contentChanged === 'changed') {
    console.log(
      `${chalk.bold('changed')} ${relativeDestPath} ${chalk.yellow(
        '[overwriting]',
      )}`,
    );
    return 'overwrite';
  }
  if (contentChanged === 'identical') {
    return 'keep';
  }
  throw new CodedError(
    'Autolinking',
    `Unknown file changed state: ${relativeDestPath}, ${contentChanged}`,
  );
}

async function upgradeFileContentChangedCallback(
  absoluteSrcFilePath: string,
  relativeDestPath: string,
  contentChanged: ContentChangedCallbackOption,
): Promise<'keep' | 'overwrite'> {
  if (contentChanged === 'new') {
    console.log(`${chalk.bold('new')} ${relativeDestPath}`);
    return 'overwrite';
  }
  if (contentChanged === 'changed') {
    console.log(
      `${chalk.bold(relativeDestPath)} ` +
        `has changed in the new version.\nDo you want to keep your ${relativeDestPath} or replace it with the ` +
        'latest version?\nMake sure you have any changes you made to this file saved somewhere.\n' +
        `You can see the new version here: ${absoluteSrcFilePath}`,
    );

    const {shouldReplace} = await prompts([
      {
        name: 'shouldReplace',
        type: 'confirm',
        message: `Do you want to replace ${relativeDestPath}?`,
        initial: false,
      },
    ]);

    return shouldReplace ? 'overwrite' : 'keep';
  }
  if (contentChanged === 'identical') {
    return 'keep';
  }
  throw new CodedError(
    'Autolinking',
    `Unknown file changed state: ${relativeDestPath}, ${contentChanged}`,
  );
}
