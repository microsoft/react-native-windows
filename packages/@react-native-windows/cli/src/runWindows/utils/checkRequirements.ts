/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 * @format
 */

import {CodedError} from '@react-native-windows/telemetry';
import {execSync} from 'child_process';
import path from 'path';
import shell from 'shelljs';
import Version from './version';

type Requirement =
  | 'os'
  | 'msbuild'
  | 'visualstudio'
  | 'windowssdk'
  | 'phonesdk';

const REQUIRED_VERSIONS: Record<string, Record<Requirement, Version>> = {
  '10.0': {
    os: new Version(6, 3),
    msbuild: new Version(14, 0),
    visualstudio: new Version(14, 0),
    windowssdk: new Version(10, 0),
    phonesdk: new Version(10, 0),
  },
};

function shortenVersion(version: Version): string {
  return /^(\d+(?:\.\d+)?)/.exec(version.toString())![1];
}

function getMinimalRequiredVersionFor(
  requirement: Requirement,
  windowsTargetVersion: keyof typeof REQUIRED_VERSIONS,
): Version {
  return REQUIRED_VERSIONS[windowsTargetVersion][requirement];
}

function getInstalledWindowsSdks(): Version[] {
  const installedSdks: Version[] = [];

  const execString =
    'reg query "HKLM\\SOFTWARE\\Microsoft\\Microsoft SDKs\\Windows" /s /v InstallationFolder /reg:32';
  let output: string;
  try {
    output = execSync(execString).toString();
  } catch (e) {
    return installedSdks;
  }

  const re = /\\Microsoft SDKs\\Windows\\v(\d+\.\d+)\s*InstallationFolder\s+REG_SZ\s+(.*)/gim;
  let match: RegExpExecArray | null;
  while ((match = re.exec(output))) {
    const sdkPath = match[2];
    if (shell.test('-e', path.join(sdkPath, 'SDKManifest.xml'))) {
      const sdkVersion = Version.tryParse(match[1]);
      if (!sdkVersion) {
        throw new CodedError(
          'BadSDKVersionFormat',
          `Unexpected SDK version format for '${match[1]}'`,
        );
      }
      installedSdks.push(sdkVersion);
    }
  }

  return installedSdks;
}

function checkWinSdk(windowsTargetVersion: string): string {
  const installedSdks = getInstalledWindowsSdks();
  const requiredVersion = getMinimalRequiredVersionFor(
    'windowssdk',
    windowsTargetVersion,
  );
  const hasSdkInstalled = installedSdks.some(installedSdk =>
    installedSdk.eq(requiredVersion),
  );

  if (hasSdkInstalled) {
    return shortenVersion(requiredVersion);
  }

  const shortenedVersion = shortenVersion(requiredVersion);
  throw new CodedError(
    'NoSDK',
    `Windows SDK not found. Ensure that you have installed Windows ${shortenedVersion} SDK along with Visual Studio or install Windows ${shortenedVersion} SDK separately from https://dev.windows.com/en-us/downloads`,
  );
}

export function isWinSdkPresent(target: string): string {
  return checkWinSdk(target);
}
