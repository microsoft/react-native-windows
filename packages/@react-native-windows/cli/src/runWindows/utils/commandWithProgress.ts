/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 * @format
 */

import {spawn, SpawnOptions} from 'child_process';
import ora from 'ora';
import spinners from 'cli-spinners';
import chalk from 'chalk';
import {CodedError, CodedErrorType} from '@react-native-windows/telemetry';

function setSpinnerText(spinner: ora.Ora, prefix: string, text: string) {
  text = prefix + spinnerString(text);
  spinner.text = text.length > 45 ? text.slice(0, 110) + '...' : text;
}

function spinnerString(msg: string) {
  msg = msg.trim();
  const lastLineIndex = msg.lastIndexOf('\n');
  if (lastLineIndex > 0) {
    msg = msg.slice(lastLineIndex + 1);
  }
  msg = msg.trim();
  return msg;
}

export function newSpinner(text: string) {
  const options: ora.Options = {
    text: text,
    indent: 1,

    // ora turns off emoji rendering by default on windows, since the default terminal
    // doesn't render them.  Turn them back on when running in the new windows terminal
    ...(process.env.WT_SESSION && {spinner: spinners.dots}),

    // By default, ora process.stderr as the output stream, however,the VS Code debugger
    // Uses stdout to match success patterns
    stream: process.stdout,
  };

  return ora(options).start();
}

export async function runPowerShellScriptFunction(
  taskDescription: string,
  script: string | null,
  funcName: string,
  verbose: boolean,
  errorCategory: CodedErrorType,
) {
  try {
    const printException = verbose ? '$_;' : '';
    const importScript = script ? `Import-Module "${script}"; ` : '';
    await commandWithProgress(
      newSpinner(taskDescription),
      taskDescription,
      'powershell',
      [
        '-NoProfile',
        '-ExecutionPolicy',
        'RemoteSigned',
        `${importScript}try { ${funcName} -ErrorAction Stop; $lec = $LASTEXITCODE; } catch { $lec = 1; ${printException} }; exit $lec`,
      ],
      verbose,
      errorCategory,
    );
  } catch {
    // The error output from the process will be shown if verbose is set.
    // We don't capture the process output if verbose is set, but at least we have the task name in text, so throw that.
    throw new CodedError(errorCategory, taskDescription);
  }
}

export function commandWithProgress(
  spinner: ora.Ora,
  taskDoingName: string,
  command: string,
  args: string[],
  verbose: boolean,
  errorCategory: CodedErrorType,
) {
  return new Promise((resolve, reject) => {
    const spawnOptions: SpawnOptions = verbose ? {stdio: 'inherit'} : {};

    if (verbose) {
      spinner.stop();
    }

    const cp = spawn(command, args, spawnOptions);
    let firstErrorLine: string | null = null;
    if (!verbose) {
      cp.stdout!.on('data', chunk => {
        const text = chunk.toString();
        setSpinnerText(spinner, taskDoingName + ': ', text);
      });
      cp.stderr!.on('data', chunk => {
        const text: string = chunk.toString();
        if (!firstErrorLine) {
          firstErrorLine = text;
        }
        if (text) {
          setSpinnerText(spinner, taskDoingName + ': ERROR: ', firstErrorLine);
        }
        reject(
          new CodedError(errorCategory, firstErrorLine, {
            taskName: taskDoingName,
          }),
        );
      });
    }
    cp.on('error', e => {
      if (verbose) {
        console.error(chalk.red(e.toString()));
      }
      spinner.fail(e.toString());
      const ce = new CodedError(errorCategory, e.message);
      ce.stack = e.stack;
      reject(ce);
    }).on('close', code => {
      if (code === 0) {
        spinner.succeed(taskDoingName);
        resolve();
      } else {
        spinner.fail();
        reject(
          new CodedError(
            errorCategory,
            `${taskDoingName} - error code ${code}`,
            {
              errorCode: code,
            },
          ),
        );
      }
    });
  });
}

export function newError(text: string) {
  newSpinner(text).fail(text);
}

export function newWarn(text: string) {
  newSpinner(text).warn(text);
}

export function newSuccess(text: string) {
  newSpinner(text).succeed(text);
}

export function newInfo(text: string) {
  newSpinner(text).info(text);
}
