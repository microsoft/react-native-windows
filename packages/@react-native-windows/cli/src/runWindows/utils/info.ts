/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 * @format
 */

// @ts-ignore no typings
import envinfo from 'envinfo';

export async function getEnvironmentInfo(): Promise<string> {
  return await envinfo.run({
    System: ['OS', 'CPU', 'Memory', 'Shell'],
    Binaries: ['Node', 'Yarn', 'npm', 'Watchman'],
    npmPackages: [
      'react',
      'react-native',
      'react-native-windows',
      '@react-native-community/cli',
      'react-native-cli',
    ],
    npmGlobalPackages: ['*react-native*'],
  });
}
