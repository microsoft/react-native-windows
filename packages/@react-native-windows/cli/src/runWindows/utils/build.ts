/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 * @format
 */

import path from 'path';

import MSBuildTools from './msbuildtools';
import Version from './version';
import {newError} from './commandWithProgress';
import {RunWindowsOptions, BuildConfig, BuildArch} from '../runWindowsOptions';
import {Config} from '@react-native-community/cli-types';
import {CodedError} from '@react-native-windows/telemetry';

export async function buildSolution(
  buildTools: MSBuildTools,
  slnFile: string,
  buildType: BuildConfig,
  buildArch: BuildArch,
  msBuildProps: Record<string, string>,
  verbose: boolean,
  target: 'build' | 'deploy',
  buildLogDirectory?: string,
  singleproc?: boolean,
) {
  const minVersion = new Version(10, 0, 19041, 0);
  const allVersions = MSBuildTools.getAllAvailableUAPVersions();
  if (!allVersions.some(v => v.gte(minVersion))) {
    throw new CodedError(
      'MinSDKVersionNotMet',
      'Must have a minimum Windows SDK version 10.0.19041.0 installed',
    );
  }

  await buildTools.buildProject(
    slnFile,
    buildType,
    buildArch,
    msBuildProps,
    verbose,
    target,
    buildLogDirectory,
    singleproc,
  );
}

const configErrorString = 'Error: ';

export function getAppSolutionFile(options: RunWindowsOptions, config: Config) {
  // Use the solution file if specified
  if (options.sln) {
    return path.join(options.root, options.sln);
  }

  // Check the answer from react-native config
  const windowsAppConfig = config.project.windows;
  if (!windowsAppConfig) {
    throw new CodedError(
      'NoWindowsConfig',
      "Couldn't determine Windows app config",
    );
  }
  const configSolutionFile = windowsAppConfig.solutionFile;

  if (configSolutionFile.startsWith(configErrorString)) {
    newError(
      configSolutionFile.substr(configErrorString.length) +
        ' Optionally, use --sln {slnFile}.',
    );
    return null;
  } else {
    return path.join(
      windowsAppConfig.folder,
      windowsAppConfig.sourceDir,
      configSolutionFile,
    );
  }
}

export function getAppProjectFile(options: RunWindowsOptions, config: Config) {
  // Use the project file if specified
  if (options.proj) {
    return path.join(options.root, options.proj);
  }

  // Check the answer from react-native config
  const windowsAppConfig = config.project.windows;
  const configProject = windowsAppConfig.project;

  if (
    typeof configProject === 'string' &&
    configProject.startsWith(configErrorString)
  ) {
    newError(
      configProject.substr(configErrorString.length) +
        ' Optionally, use --proj {projFile}.',
    );
    return null;
  } else {
    const configProjectFile = configProject.projectFile;
    if (configProjectFile.startsWith(configErrorString)) {
      newError(
        configProjectFile.substr(configErrorString.length) +
          ' Optionally, use --proj {projFile}.',
      );
      return null;
    }
    return path.join(
      windowsAppConfig.folder,
      windowsAppConfig.sourceDir,
      configProjectFile,
    );
  }
}

export function parseMsBuildProps(
  options: RunWindowsOptions,
): Record<string, string> {
  const result: Record<string, string> = {};
  if (options.msbuildprops) {
    const props = options.msbuildprops.split(',');
    for (const prop of props) {
      const propAssignment = prop.split('=');
      result[propAssignment[0]] = propAssignment[1];
    }
  }
  return result;
}
