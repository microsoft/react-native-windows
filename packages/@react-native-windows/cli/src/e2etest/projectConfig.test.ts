/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 * @format
 */

import fs from 'fs';
import path from 'path';

import {
  projectConfigWindows,
  WindowsProjectConfig,
} from '../config/projectConfig';

import {copyAndReplace} from '../generator-common';
import {AutolinkWindows} from '../runWindows/utils/autolink';
import {
  ensureWinUI3Project,
  templateRoot,
  testProjectGuid,
  tryMkdir,
} from './projectConfig.utils';

type TargetProject = [string, ((folder: string) => Promise<void>)?];

function project(
  name: string,
  setup?: (folder: string) => Promise<void>,
): TargetProject {
  return [name, setup];
}

const projects: TargetProject[] = [
  // Nothing but a react-native.config.js, with nulls
  project('BlankApp'),
  // Nothing but a windows folder
  project('MissingProjectFilesApp', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);
  }),
  // New C++ project based on the template
  project('SimpleCppApp', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);

    const replacements = {
      name: 'SimpleCppApp',
      namespace: 'SimpleCppApp',
      useMustache: true,
      projectGuidUpper: testProjectGuid,
      projectGuidLower: testProjectGuid.toLowerCase(),
    };

    await copyAndReplace(
      path.join(templateRoot, 'cpp-app/proj/MyApp.sln'),
      path.join(windowsDir, 'SimpleCppApp.sln'),
      replacements,
      null,
    );

    const projDir = path.join(windowsDir, 'SimpleCppApp');
    await tryMkdir(projDir);

    await copyAndReplace(
      path.join(templateRoot, 'cpp-app/proj/MyApp.vcxproj'),
      path.join(projDir, 'SimpleCppApp.vcxproj'),
      replacements,
      null,
    );
  }),
  // New C# project based on the template
  project('SimpleCSharpApp', async (folder: string) => {
    const windowsDir = path.join(folder, 'windows');
    await tryMkdir(windowsDir);

    const replacements = {
      name: 'SimpleCSharpApp',
      namespace: 'SimpleCSharpApp',
      useMustache: true,
      projectGuidUpper: testProjectGuid,
      projectGuidLower: testProjectGuid.toLowerCase(),
    };

    await copyAndReplace(
      path.join(templateRoot, 'cs-app/proj/MyApp.sln'),
      path.join(windowsDir, 'SimpleCSharpApp.sln'),
      replacements,
      null,
    );

    const projDir = path.join(windowsDir, 'SimpleCSharpApp');
    await tryMkdir(projDir);

    await copyAndReplace(
      path.join(templateRoot, 'cs-app/proj/MyApp.csproj'),
      path.join(projDir, 'SimpleCSharpApp.csproj'),
      replacements,
      null,
    );
  }),
  project('WithWinUI3', async (folder: string) => {
    await ensureWinUI3Project(folder);
  }),
  project('WithIndirectDependency'),
];

// Tests that given userConfig is null, the result will always be null
test.each(projects)(
  'projectConfig - %s (userConfig is null)',
  async (name, setup) => {
    const folder = path.resolve('src/e2etest/projects/', name);

    if (setup !== undefined) {
      await setup(folder);
    }

    const userConfig = null;

    expect(projectConfigWindows(folder, userConfig)).toBeNull();
  },
);

// Tests the result given a windows project config in react-native.config.js
test.each(projects)(
  'projectConfig - %s (Use react-native.config.js)',
  async (name, setup) => {
    const folder = path.resolve('src/e2etest/projects/', name);
    const rnc = require(path.join(folder, 'react-native.config.js'));

    if (setup !== undefined) {
      await setup(folder);
    }

    const userConfig: Partial<WindowsProjectConfig> = rnc.project.windows;

    if (name === 'BlankApp') {
      expect(projectConfigWindows(folder, userConfig)).toMatchSnapshot();
    } else {
      expect(projectConfigWindows(folder, userConfig)).toMatchSnapshot({
        folder: expect.stringContaining(name),
      });
    }
  },
);

// Tests the result of ignoring the windows project config in react-native.config.js
test.each(projects)(
  'projectConfig - %s (Ignore react-native.config.js)',
  async (name, setup) => {
    const folder = path.resolve('src/e2etest/projects/', name);

    if (setup !== undefined) {
      await setup(folder);
    }

    const userConfig: Partial<WindowsProjectConfig> = {};

    if (name === 'BlankApp') {
      expect(projectConfigWindows(folder, userConfig)).toMatchSnapshot();
    } else {
      expect(projectConfigWindows(folder, userConfig)).toMatchSnapshot({
        folder: expect.stringContaining(name),
      });
    }
  },
);

test('useWinUI3=true in react-native.config.js, useWinUI3=false in ExperimentalFeatures.props', async () => {
  const folder = path.resolve('src/e2etest/projects/WithWinUI3');

  const rnc = require(path.join(folder, 'react-native.config.js'));

  const config = projectConfigWindows(folder, rnc.project.windows)!;

  const al = new AutolinkWindows(
    {windows: config},
    {},
    {
      check: false,
      logging: false,
    },
  );

  const exd = await al.ensureXAMLDialect();
  expect(exd).toBeTruthy();

  const packagesConfig = (
    await fs.promises.readFile(
      path.join(folder, 'windows/WithWinUI3/packages.config'),
    )
  ).toString();

  const experimentalFeatures = (
    await fs.promises.readFile(
      path.join(folder, 'windows/ExperimentalFeatures.props'),
    )
  ).toString();

  expect(packagesConfig.replace(/\r/g, '')).toMatchSnapshot();

  expect(experimentalFeatures.replace(/\r/g, '')).toMatchSnapshot();
});
