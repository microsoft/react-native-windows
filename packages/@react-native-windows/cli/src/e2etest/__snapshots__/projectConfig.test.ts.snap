// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectConfig - BlankApp (Ignore react-native.config.js) 1`] = `null`;

exports[`projectConfig - BlankApp (Use react-native.config.js) 1`] = `null`;

exports[`projectConfig - MissingProjectFilesApp (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "MissingProjectFilesApp",
  "project": Object {
    "projectFile": "Error: No app project file found, please specify in react-native.config.",
  },
  "solutionFile": "Error: No app solution file found, please specify in react-native.config.",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - MissingProjectFilesApp (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "MissingProjectFilesApp",
  "project": Object {
    "projectFile": "Error: Project is required but not specified in react-native.config.",
  },
  "solutionFile": "Error: Solution file is required but not specified in react-native.config.",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - SimpleCSharpApp (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCSharpApp",
  "project": Object {
    "projectFile": "SimpleCSharpApp\\\\SimpleCSharpApp.csproj",
    "projectGuid": "{416476D5-974A-4EE2-8145-4E331297247E}",
    "projectLang": "cs",
    "projectName": "SimpleCSharpApp",
  },
  "solutionFile": "SimpleCSharpApp.sln",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - SimpleCSharpApp (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCSharpApp",
  "project": Object {
    "projectFile": "SimpleCSharpApp\\\\SimpleCSharpApp.csproj",
    "projectGuid": "{416476D5-974A-4EE2-8145-4E331297247E}",
    "projectLang": "cs",
    "projectName": "SimpleCSharpApp",
  },
  "solutionFile": "SimpleCSharpApp.sln",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - SimpleCppApp (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCppApp",
  "project": Object {
    "projectFile": "SimpleCppApp\\\\SimpleCppApp.vcxproj",
    "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
    "projectLang": "cpp",
    "projectName": "SimpleCppApp",
  },
  "solutionFile": "SimpleCppApp.sln",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - SimpleCppApp (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "SimpleCppApp",
  "project": Object {
    "projectFile": "SimpleCppApp\\\\SimpleCppApp.vcxproj",
    "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
    "projectLang": "cpp",
    "projectName": "SimpleCppApp",
  },
  "solutionFile": "SimpleCppApp.sln",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - WithIndirectDependency (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "WithIndirectDependency",
  "project": Object {
    "projectFile": "Error: Too many app project files found, please specify in react-native.config.",
  },
  "solutionFile": "WithIndirectDependency.sln",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - WithIndirectDependency (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "WithIndirectDependency",
  "project": Object {
    "projectFile": "WithIndirectDependency\\\\WithIndirectDependency.vcxproj",
    "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
    "projectLang": "cpp",
    "projectName": "WithIndirectDependency",
  },
  "solutionFile": "WithIndirectDependency.sln",
  "sourceDir": "windows",
  "useWinUI3": false,
}
`;

exports[`projectConfig - WithWinUI3 (Ignore react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "WithWinUI3",
  "project": Object {
    "projectFile": "WithWinUI3\\\\WithWinUI3.vcxproj",
    "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
    "projectLang": "cpp",
    "projectName": "WithWinUI3",
  },
  "solutionFile": "WithWinUI3.sln",
  "sourceDir": "windows",
}
`;

exports[`projectConfig - WithWinUI3 (Use react-native.config.js) 1`] = `
Object {
  "folder": StringContaining "WithWinUI3",
  "project": Object {
    "projectFile": "WithWinUI3\\\\WithWinUI3.vcxproj",
    "projectGuid": "{416476d5-974a-4ee2-8145-4e331297247e}",
    "projectLang": "cpp",
    "projectName": "WithWinUI3",
  },
  "solutionFile": "WithWinUI3.sln",
  "sourceDir": "windows",
  "useWinUI3": true,
}
`;

exports[`useWinUI3=true in react-native.config.js, useWinUI3=false in ExperimentalFeatures.props 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<packages>
  <package id=\\"Microsoft.ReactNative.Cxx\\" version=\\"1.0.0\\" targetFramework=\\"native\\"/>
  <package id=\\"Microsoft.WinUI\\" version=\\"3.0.0-preview4.210210.4\\" targetFramework=\\"native\\"/>
</packages>"
`;

exports[`useWinUI3=true in react-native.config.js, useWinUI3=false in ExperimentalFeatures.props 2`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<Project xmlns=\\"http://schemas.microsoft.com/developer/msbuild/2003\\">

  <!--
    This file contains some important settings that will apply globally for
    your app and *all* native modules your app consumes. These values were
    set when you created the app project, and in some cases cannot be
    simply changed here without recreating a new project.
  -->

  <PropertyGroup Label=\\"Microsoft.ReactNative Experimental Features\\">
    <!--
      Enables default usage of Hermes.
      
      See https://microsoft.github.io/react-native-windows/docs/hermes
    -->
    <UseHermes>false</UseHermes>

    <!--
      Changes compilation to assume use of WinUI 3 instead of System XAML.
      Requires creation of new project.

      See https://microsoft.github.io/react-native-windows/docs/winui3
    -->
    <UseWinUI3>true</UseWinUI3>

    <!--
      Changes compilation to assume use of Micorsoft.ReactNative NuGets packages instead of source.
      Requires creation of new project.
    -->
    <UseExperimentalNuget>false</UseExperimentalNuget>
  
  </PropertyGroup>

</Project>"
`;
